File: frontend/__mocks__/metadata.ts
/**
 * What a section declares, and what its documents actually recorded.
 *
 * Three surfaces read the same two shapes — the metadata block on a document,
 * the section table, and the field editor in settings — so they are written
 * once here. A story that invents its own spec is how the block and the table
 * end up demonstrating different rules.
 *
 * The corpus is the one from issue #79 rather than the catalogue's usual Vue
 * material: a section of eighteen sibling specifications, in Russian, where the
 * same five facts were being typed into the first five lines of prose. Two
 * things only show up in that shape — a long Cyrillic label against
 * `--metadata-col-max`, and a field that every document leaves blank.
 */
import type { FieldSpec } from '~/composables/useFieldSpec'

/**
 * The declaration on "Спецификации".
 *
 * `stage` is an enum on purpose and the issue's whole argument for types is in
 * it: narrowing the answer space is what raises fill rate, while a typed date
 * fills no better than an untyped one. It sits beside the system `status`,
 * which is the pair the block is built to show — they are the two that
 * disagreed in prose.
 */
export const specSpecifications: FieldSpec[] = [
  { key: 'stage', label: 'Стадия', type: 'enum', options: ['draft', 'in-review', 'agreed', 'superseded'] },
  { key: 'produces', label: 'Produces', type: 'text', repeated: true },
  { key: 'consumes', label: 'Consumes', type: 'text', repeated: true },
  {
    key: 'owner',
    label: 'Owner',
    type: 'text',
    required: true,
    help: 'The service name from the repo. Ask rather than guess.',
  },
  { key: 'registry', label: 'Registry', type: 'ref' },
  { key: 'reviewed', label: 'Reviewed', type: 'date' },
]

/**
 * One declared field. The common shape after somebody's first edit, and the
 * case the table deliberately refuses to offer a toggle for.
 */
export const specSingleField: FieldSpec[] = [
  { key: 'stage', label: 'Стадия', type: 'enum', options: ['draft', 'in-review', 'agreed'] },
]

/**
 * Twelve fields — `MaxSectionFields` — with five required, which is the second
 * and harder cap (`MaxRequiredFields`). They are independent, so a section can
 * sit on both at once and this is what that looks like.
 */
export const specAtCap: FieldSpec[] = [
  ...specSpecifications,
  { key: 'component', label: 'Component', type: 'text', required: true, help: 'The deployable this spec belongs to.' },
  { key: 'transport', label: 'Transport', type: 'enum', options: ['http', 'grpc', 'queue', 'temporal'], required: true },
  { key: 'schema_url', label: 'Schema', type: 'url' },
  { key: 'retries', label: 'Retries', type: 'number' },
  { key: 'supersedes', label: 'Supersedes', type: 'ref', repeated: true },
  { key: 'reviewer', label: 'Reviewer', type: 'text', required: true, help: 'Whoever signed off. Ask if unclear.' },
]

/**
 * Labels and values as this product actually receives them. `--metadata-col-max:
 * 16rem` is the only thing between the first label here and a value column
 * pushed off the page.
 */
export const specCyrillic: FieldSpec[] = [
  { key: 'stage', label: 'Стадия согласования спецификации площадки', type: 'text' },
  { key: 'produces', label: 'Кто формирует полезную нагрузку', type: 'text', repeated: true },
  {
    key: 'owner',
    label: 'Ответственная команда сопровождения',
    type: 'text',
    required: true,
    help: 'Название сервиса из репозитория. Спросите, если неясно, а не угадывайте.',
  },
]

/** Every declared field answered. */
export const metadataFilled: Record<string, unknown> = {
  stage: 'in-review',
  produces: ['scanner-watchdog'],
  consumes: ['scanner-orchestrator'],
  owner: 'platform',
  registry: 'E47',
  reviewed: '2026-08-14',
}

/** Two answered, the required one not. The ordinary in-progress document. */
export const metadataPartial: Record<string, unknown> = {
  stage: 'draft',
  produces: ['scanner-watchdog'],
}

/**
 * Nothing answered at all — a document created before the section declared
 * anything, or written by an agent that ignored the schema. Every row is the
 * blank cell the feature exists to make visible.
 */
export const metadataNothingFilled: Record<string, unknown> = {}

/**
 * The explicit unknown, which is `null` and not absence. It answers a required
 * field without inventing a fact, which is the counterweight to an author that
 * never leaves a blank.
 */
export const metadataUnknownOwner: Record<string, unknown> = {
  ...metadataFilled,
  owner: null,
}

/** A stored value that no longer matches its declaration. Kept, flagged, re-checked on read. */
export const metadataInvalidStage: Record<string, unknown> = {
  ...metadataFilled,
  stage: 'sent for review',
}

/**
 * Values under keys nothing declares any more — the section was rewritten and
 * every old key was dropped. Removing a field decides what gets collected next;
 * it is not a verdict on what was already recorded, so all of this still shows.
 */
export const metadataAllOrphaned: Record<string, unknown> = {
  implemented_by: 'scanner-watchdog, scanner-orchestrator',
  related: 'SPEC-02, SPEC-03, SPEC-04',
  reviewed_on: '14.08.2026',
  status_note: 'черновик на ревью',
}

/** Cyrillic values against the Cyrillic labels. */
export const metadataCyrillic: Record<string, unknown> = {
  stage: 'черновик, отправленный на ревью команде платформы и оркестратора',
  produces: ['scanner-watchdog', 'scanner-orchestrator', 'cluster-watchdog'],
  owner: 'команда платформы наблюдаемости площадок',
}

/**
 * The shape a section instruction is supposed to have: three to six
 * imperatives, naming the section's own declared fields.
 *
 * Not invented — this is the convention the eighteen hand-written SPEC
 * preambles in R21 were all groping towards, written once above the documents
 * instead of eighteen times inside them.
 */
export const mockSectionInstruction =
  'Назови производящий сервис в поле service.\n' +
  'Назови потребителя в поле consumer.\n' +
  'Один абзац обоснования, затем полезная нагрузка.\n' +
  'Заполни owner и status до сохранения.'

/**
 * The section itself, shaped as the research payload delivers it — `field_spec`,
 * `spec_version` and `instruction` beside the ordinary section fields.
 *
 * It carries an instruction, so every story that hands this to `EntriesView`
 * shows the read block above the documents as well as the field chips.
 */
export const mockSpecSection = {
  id: 'sec_spec',
  code: 'S13',
  name: 'specifications',
  display_name: 'Спецификации',
  description: 'Одна спецификация на блок реестра.',
  status: 'active',
  entries_count: 6,
  field_spec: specSpecifications,
  spec_version: 4,
  instruction: mockSectionInstruction,
}

/** A section that declares nothing and says nothing about how to write in it:
 *  a topic, not a class of document — which is most sections. */
export const mockTopicSection = {
  id: 'sec_questions',
  code: 'S11',
  name: 'open-questions',
  display_name: 'Вопросы на повестку',
  description: null,
  status: 'active',
  entries_count: 8,
  field_spec: [],
  spec_version: 0,
  instruction: '',
}

/**
 * Six sibling specifications for the table.
 *
 * `reviewed` is empty on every one of them, deliberately. That column is the
 * output the feature exists to produce: a field nobody fills is not noise, it
 * is the answer to whether the field should have been declared — and it is only
 * legible as a column of dashes standing next to filled ones.
 */
export const mockSpecEntries = [
  {
    id: 'ent_050', code: 'E50', section_id: 'sec_spec',
    title: 'SPEC-01 · Payload состояния площадки',
    description: 'Первая спецификация блока 1 по реестру [[E47]].',
    status: 'active', tags: ['spec', 'payload'], spec_version: 4,
    metadata: { stage: 'in-review', produces: ['scanner-watchdog'], consumes: ['scanner-orchestrator'], owner: 'platform', registry: 'E47' },
  },
  {
    id: 'ent_051', code: 'E51', section_id: 'sec_spec',
    title: 'SPEC-02 · Пробы площадки и вердикт о её работоспособности',
    description: 'Что считается пробой и кто выносит вердикт.',
    status: 'active', tags: ['spec', 'probes'], spec_version: 4,
    metadata: { stage: 'agreed', produces: ['scanner-probe'], consumes: ['scanner-orchestrator', 'temporal-watchdog'], owner: 'platform', registry: 'E47' },
  },
  {
    id: 'ent_052', code: 'E52', section_id: 'sec_spec',
    title: 'SPEC-03 · Сервис watchdog',
    description: 'Границы ответственности сторожа.',
    status: 'completed', tags: ['spec'], spec_version: 4,
    // The required field unanswered: one "1 missing" chip, one blank cell.
    metadata: { stage: 'agreed', produces: ['scanner-watchdog'] },
  },
  {
    id: 'ent_053', code: 'E53', section_id: 'sec_spec',
    title: 'SPEC-04 · platform-profile',
    description: 'Профиль площадки как единица конфигурации.',
    status: 'draft', tags: ['spec', 'config'], spec_version: 3,
    // Explicit unknown: the table prints "unknown", not a dash, and the chip
    // stays away — somebody looked, which is a different fact from silence.
    metadata: { stage: 'draft', owner: null, registry: 'E47' },
  },
  {
    id: 'ent_054', code: 'E54', section_id: 'sec_spec',
    title: 'SPEC-05 · Инциденты без права решения',
    description: 'Эмиттер, который не принимает решений.',
    status: 'draft', tags: ['spec'], spec_version: 4,
    // Nothing at all. Written before the section declared anything.
    metadata: {},
  },
  {
    id: 'ent_055', code: 'E55', section_id: 'sec_spec',
    title: 'SPEC-06 · Реестр блоков и его владельцы',
    description: 'Кто ведёт реестр и по каким правилам.',
    status: 'active', tags: ['spec', 'registry'], spec_version: 4,
    metadata: {
      stage: 'superseded',
      produces: ['registry-api', 'registry-worker', 'registry-cli', 'registry-exporter'],
      consumes: ['scanner-orchestrator'],
      owner: 'команда платформы наблюдаемости площадок',
      registry: 'E47',
    },
  },
]

/**
 * The rules, as `GET /api/metadata/schema` returns them.
 *
 * Copied from `domain.FieldSchema()` rather than invented: the editor reads
 * them from the server precisely so a cap it believes and a cap the server
 * enforces cannot disagree, and a mock that made up its own numbers would put
 * the disagreement back into the catalogue.
 */
export const fieldCaps = {
  fields: 12,
  required: 5,
  options: 20,
  values: 20,
  text_max: 200,
  label_max: 60,
  help_max: 200,
  key_max: 32,
  key_pattern: '^[a-z][a-z0-9_]*$',
  instruction_max: 500,
}

export const fieldTypes = [
  { type: 'enum', needs_option: true, repeatable: true, hint: 'A short list of named options.' },
  { type: 'ref', repeatable: true, hint: 'A short code of something in this product: E12, R2:E5, RM1.' },
  { type: 'date', repeatable: false, hint: 'A calendar date, YYYY-MM-DD.' },
  { type: 'text', repeatable: true, hint: 'One short line.' },
  { type: 'number', repeatable: false, hint: 'A number.' },
  { type: 'url', repeatable: true, hint: 'An http or https address.' },
]

/** The keys the Obsidian export already emits. Sorted, as the server sorts them. */
export const reservedKeys = [
  'aliases', 'code', 'created', 'research', 'section',
  'session', 'status', 'tags', 'title', 'type', 'updated',
]

File: internal/domain/metadata.go
package domain

import (
	"fmt"
	"net/url"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"
	"unicode/utf8"
)

// A section declares the fields its documents record; an entry carries the
// values. The vocabulary is closed — only a declared key may be written — which
// is what makes a column-per-field view over a section possible, and what lets
// a section that declares nothing behave as it did before the feature existed.
//
// The division of labour, which every later change should be checked against:
// `tags` are free-form and cross-cutting, a metadata field is a typed slot in a
// per-section schema. They answer different questions and neither replaces the
// other.

type FieldType string

const (
	// FieldEnum is a controlled vocabulary. It is the type that actually moves
	// fill rates: a field converted from free text to four to six named options
	// fills several times better, while a typed date or text field fills no
	// better than an untyped one. Types help exactly insofar as they narrow the
	// answer space, so prefer this wherever a field can bear it.
	FieldEnum FieldType = "enum"
	// FieldRef is one or more short codes — E47, R2:E5. Stored as written.
	FieldRef  FieldType = "ref"
	FieldDate FieldType = "date"
	// FieldText is a single line, capped hard. See MaxFieldValueLen.
	FieldText   FieldType = "text"
	FieldNumber FieldType = "number"
	FieldURL    FieldType = "url"
)

// Valid reports whether a caller may declare this type.
//
// The list is closed, and extending it is a code change rather than a
// configuration one, because of the rule that keeps this from becoming a second
// content system: a metadata field may only hold a value you could put in a
// WHERE clause or a facet. No rich text, no long text. If a value needs
// formatting, a paragraph break, or a reference inside prose, it is content and
// belongs in the document.
//
// `boolean` is deliberately absent: absent and false are indistinguishable to a
// reader, so a two-value enum with real words says more.
func (t FieldType) Valid() bool {
	switch t {
	case FieldEnum, FieldRef, FieldDate, FieldText, FieldNumber, FieldURL:
		return true
	}
	return false
}

const (
	// MaxSectionFields protects the completeness signal and the legibility of
	// the block, not the database. Past a dozen the block stops being scannable
	// at the top of a document and nobody reads it — which returns you to the
	// failure this feature exists to fix, where the reader believes the prose
	// instead. Just past the cap people start encoding structure in field names
	// (contact_1, contact_2) and metadata begins carrying content.
	MaxSectionFields = 12
	// MaxRequiredFields is a separate and harder ceiling, because required is
	// the thing that turns a schema edit into non-compliance across every
	// document at once — and because a model, unlike a person, does not leave a
	// field blank when it does not know. It fills it plausibly. Blank used to
	// mean "unknown"; with an agent author there are no blanks.
	MaxRequiredFields = 5
	MaxEnumOptions    = 20
	// MaxFieldValueLen is counted in runes, so a Cyrillic value is not worth
	// half a Latin one.
	MaxFieldValueLen = 200
	MaxFieldLabelLen = 60
	MaxFieldHelpLen  = 200
	MaxFieldKeyLen   = 32
	// MaxImportFileBytes caps one dropped markdown file. Generous next to any
	// document somebody wrote by hand, and small enough that a mis-drop of a
	// database dump is refused before it is read into memory.
	MaxImportFileBytes = 1 << 20
	// MaxRepeatedValues bounds a repeated field so one entry cannot make the
	// section table unreadable on its own.
	MaxRepeatedValues = 20
)

var fieldKeyPattern = regexp.MustCompile(`^[a-z][a-z0-9_]*$`)

// reservedFieldKeys are the keys the Obsidian export already emits as YAML
// front matter. `frontmatter.add` appends blindly and the renderer emits
// everything, so YAML last-wins: a user field keyed `status` would silently
// overwrite the system value in every exported note, and a collision on `code`
// or `aliases` would break every [[E3]] in the vault, since that is what
// Obsidian link resolution hangs on.
//
// The refusal happens when the spec is defined, not when a note is rendered, so
// the error reaches the person who made the mistake instead of surfacing in an
// export months later.
var reservedFieldKeys = map[string]bool{
	"code": true, "title": true, "aliases": true, "research": true,
	"section": true, "type": true, "status": true, "tags": true,
	"created": true, "updated": true, "session": true,
}

// ReservedFieldKeys lists the refused keys, for an error message and for the UI.
func ReservedFieldKeys() []string {
	out := make([]string, 0, len(reservedFieldKeys))
	for k := range reservedFieldKeys {
		out = append(out, k)
	}
	sortStrings(out)
	return out
}

// FieldSpec is one declared field.
//
// Key and Label are separate on purpose: the key is an immutable identifier and
// the label is what gets renamed. A "rename" that changes the key is a removal
// plus an addition and has to be spelled that way, or the values recorded under
// the old key become unreachable without anyone deciding that they should.
type FieldSpec struct {
	Key      string    `json:"key"`
	Label    string    `json:"label"`
	Type     FieldType `json:"type"`
	Required bool      `json:"required"`
	Repeated bool      `json:"repeated,omitempty"`
	// Options is the vocabulary for FieldEnum, ignored otherwise.
	Options []string `json:"options,omitempty"`
	// Help says where the value comes from — "the service name from the repo;
	// ask if unclear". The agent reads it, which is the whole reason it exists:
	// a required field with no provenance note is an invitation to invent.
	Help string `json:"help,omitempty"`
}

// ValidateFieldSpecs checks a section's whole declaration and returns one
// message per problem. It returns errors rather than a single failure because
// the caller is editing a form and deserves to see every mistake at once.
func ValidateFieldSpecs(specs []FieldSpec) []string {
	var errs []string
	if len(specs) > MaxSectionFields {
		errs = append(errs, fmt.Sprintf("a section may declare at most %d fields, got %d", MaxSectionFields, len(specs)))
	}

	seen := map[string]bool{}
	required := 0
	for i, f := range specs {
		where := fmt.Sprintf("fields[%d]", i)

		switch {
		case f.Key == "":
			errs = append(errs, where+": key is required")
		case !fieldKeyPattern.MatchString(f.Key):
			errs = append(errs, fmt.Sprintf("%s: key %q must be lowercase, start with a letter, and hold only letters, digits and underscores", where, f.Key))
		case utf8.RuneCountInString(f.Key) > MaxFieldKeyLen:
			errs = append(errs, fmt.Sprintf("%s: key %q is longer than %d characters", where, f.Key, MaxFieldKeyLen))
		case reservedFieldKeys[f.Key]:
			errs = append(errs, fmt.Sprintf("%s: key %q is reserved — the export already emits it as front matter, and a second one would overwrite it", where, f.Key))
		case seen[f.Key]:
			errs = append(errs, fmt.Sprintf("%s: key %q is declared twice", where, f.Key))
		default:
			seen[f.Key] = true
		}

		if !f.Type.Valid() {
			errs = append(errs, fmt.Sprintf("%s: unknown type %q", where, f.Type))
		}
		if utf8.RuneCountInString(f.Label) > MaxFieldLabelLen {
			errs = append(errs, fmt.Sprintf("%s: label is longer than %d characters", where, MaxFieldLabelLen))
		}
		if utf8.RuneCountInString(f.Help) > MaxFieldHelpLen {
			errs = append(errs, fmt.Sprintf("%s: help is longer than %d characters", where, MaxFieldHelpLen))
		}

		if f.Type == FieldEnum {
			if len(f.Options) == 0 {
				errs = append(errs, where+": an enum field needs at least one option")
			}
			if len(f.Options) > MaxEnumOptions {
				errs = append(errs, fmt.Sprintf("%s: at most %d options, got %d", where, MaxEnumOptions, len(f.Options)))
			}
			opts := map[string]bool{}
			for _, o := range f.Options {
				if strings.TrimSpace(o) == "" {
					errs = append(errs, where+": an option may not be blank")
					continue
				}
				if opts[o] {
					errs = append(errs, fmt.Sprintf("%s: option %q is listed twice", where, o))
				}
				opts[o] = true
			}
		} else if len(f.Options) > 0 {
			errs = append(errs, fmt.Sprintf("%s: only an enum field takes options", where))
		}

		if f.Required {
			required++
		}
	}

	if required > MaxRequiredFields {
		errs = append(errs, fmt.Sprintf("at most %d fields may be required, got %d — a required field a writer cannot answer gets invented rather than left blank", MaxRequiredFields, required))
	}
	return errs
}

// MetadataIssue names one key the write could not honour as given.
type MetadataIssue struct {
	Key string `json:"key"`
	// Value is what the writer actually sent, when a caller fills it in.
	//
	// Set by the markdown import and by nothing else. An agent that sent a bad
	// value still has it; a person looking at a file they did not write does
	// not, and "stage is not one of the options" without saying what the file
	// said is a report they have to go and check by hand.
	Value  string `json:"value,omitempty"`
	Reason string `json:"reason"`
}

// MetadataReport is what a write tells its author about the metadata it was
// handed. It is attached to a response and never stored — the same shape and
// the same reason as BlockSaveReport: an author whose value was dropped has to
// be told, because the loss is otherwise silent.
//
// A write is never refused because of metadata. The author is a model in the
// middle of an interview, and a rejection there destroys the answers a person
// has already given.
type MetadataReport struct {
	Stored []string `json:"stored,omitempty"`
	// UnknownKeys were not declared by this section and are dropped. This is
	// what "closed vocabulary" means in practice.
	UnknownKeys []MetadataIssue `json:"unknown_keys,omitempty"`
	// InvalidValues were declared but did not match their type. They are stored
	// verbatim and flagged: discarding a person's value to protect a type is the
	// same mistake as refusing the write.
	InvalidValues   []MetadataIssue `json:"invalid_values,omitempty"`
	MissingRequired []string        `json:"missing_required,omitempty"`
	SpecVersion     int             `json:"spec_version"`
}

// Complete reports whether every required field carries a value.
func (r MetadataReport) Complete() bool { return len(r.MissingRequired) == 0 }

// ValidateMetadata checks submitted values against a section's declaration and
// returns what to store alongside a report of everything it could not honour.
//
// It never returns an error: nothing here is worth failing a write over.
func ValidateMetadata(specs []FieldSpec, values map[string]any) (map[string]any, MetadataReport) {
	stored := map[string]any{}
	var report MetadataReport

	byKey := map[string]FieldSpec{}
	order := make([]string, 0, len(specs))
	for _, f := range specs {
		byKey[f.Key] = f
		order = append(order, f.Key)
	}

	keys := make([]string, 0, len(values))
	for k := range values {
		keys = append(keys, k)
	}
	sortStrings(keys)

	for _, k := range keys {
		v := values[k]
		spec, declared := byKey[k]
		if !declared {
			report.UnknownKeys = append(report.UnknownKeys, MetadataIssue{
				Key:    k,
				Reason: "this section does not declare a field with this key",
			})
			continue
		}
		if v == nil {
			// A null is an explicit "we do not know", and it is the answer to a
			// required field that nobody can honestly fill. Keeping it separable
			// from an absent key is the whole point: a model, unlike a person,
			// does not leave a field blank when it does not know — it fills it
			// plausibly — so the writable unknown is what stops a required field
			// from manufacturing a fact.
			stored[k] = nil
			report.Stored = append(report.Stored, k)
			continue
		}
		if isBlankValue(v) {
			// An empty string clears the field rather than storing something that
			// would read as an answer.
			continue
		}
		normalized, err := normalizeValue(spec, v)
		if err != nil {
			// Stored anyway, and flagged.
			stored[k] = v
			report.InvalidValues = append(report.InvalidValues, MetadataIssue{Key: k, Reason: err.Error()})
			report.Stored = append(report.Stored, k)
			continue
		}
		stored[k] = normalized
		report.Stored = append(report.Stored, k)
	}

	for _, k := range order {
		f := byKey[k]
		if f.Required {
			if _, ok := stored[k]; !ok {
				report.MissingRequired = append(report.MissingRequired, k)
			}
		}
	}
	return stored, report
}

// OrphanedKeys are values a document still carries under keys the section no
// longer declares. They are never deleted — removing a field is a decision
// about future collection, not a verdict on what was already written — so the
// UI shows them apart rather than pretending they are gone.
func OrphanedKeys(specs []FieldSpec, values map[string]any) []string {
	declared := map[string]bool{}
	for _, f := range specs {
		declared[f.Key] = true
	}
	var out []string
	for k := range values {
		if !declared[k] {
			out = append(out, k)
		}
	}
	sortStrings(out)
	return out
}

// MissingRequired lists required keys a stored document does not answer. It is
// computed on read against the current spec and never persisted: a stored
// completeness flag goes stale the moment the declaration changes.
func MissingRequired(specs []FieldSpec, values map[string]any) []string {
	var out []string
	for _, f := range specs {
		if !f.Required {
			continue
		}
		v, ok := values[f.Key]
		if !ok {
			out = append(out, f.Key)
			continue
		}
		// An explicit unknown answers the requirement. It says somebody looked.
		if v != nil && isBlankValue(v) {
			out = append(out, f.Key)
		}
	}
	return out
}

func isBlankValue(v any) bool {
	switch t := v.(type) {
	case nil:
		return true
	case string:
		return strings.TrimSpace(t) == ""
	case []any:
		return len(t) == 0
	case []string:
		return len(t) == 0
	}
	return false
}

// normalizeValue coerces one submitted value into the shape its type promises,
// or explains why it cannot.
func normalizeValue(spec FieldSpec, v any) (any, error) {
	if spec.Repeated {
		items, err := toList(v)
		if err != nil {
			return nil, err
		}
		if len(items) > MaxRepeatedValues {
			return nil, fmt.Errorf("at most %d values, got %d", MaxRepeatedValues, len(items))
		}
		out := make([]any, 0, len(items))
		var firstErr error
		for _, item := range items {
			nv, err := normalizeScalar(spec, item)
			if err != nil && firstErr == nil {
				firstErr = err
			}
			out = append(out, nv)
		}
		return out, firstErr
	}
	if _, isList := v.([]any); isList {
		return nil, fmt.Errorf("field %q takes a single value, not a list", spec.Key)
	}
	return normalizeScalar(spec, v)
}

func toList(v any) ([]any, error) {
	switch t := v.(type) {
	case []any:
		return t, nil
	case []string:
		out := make([]any, len(t))
		for i, s := range t {
			out[i] = s
		}
		return out, nil
	default:
		// A single value for a repeated field is a courtesy, not an error: an
		// author writing one service name should not have to know it is a list.
		return []any{v}, nil
	}
}

func normalizeScalar(spec FieldSpec, v any) (any, error) {
	switch spec.Type {
	case FieldNumber:
		switch n := v.(type) {
		case float64:
			return n, nil
		case int:
			return float64(n), nil
		case string:
			f, err := strconv.ParseFloat(strings.TrimSpace(n), 64)
			if err != nil {
				return v, fmt.Errorf("%q is not a number", n)
			}
			return f, nil
		default:
			return v, fmt.Errorf("expected a number")
		}
	}

	s, ok := v.(string)
	if !ok {
		return v, fmt.Errorf("expected text")
	}
	s = strings.TrimSpace(oneLineValue(s))
	if utf8.RuneCountInString(s) > MaxFieldValueLen {
		return v, fmt.Errorf("longer than %d characters — a value that needs more room is content, and belongs in the document", MaxFieldValueLen)
	}

	switch spec.Type {
	case FieldEnum:
		for _, o := range spec.Options {
			if o == s {
				return s, nil
			}
		}
		return s, fmt.Errorf("%q is not one of: %s", s, strings.Join(spec.Options, ", "))
	case FieldDate:
		if _, err := time.Parse("2006-01-02", s); err == nil {
			return s, nil
		}
		if t, err := time.Parse(time.RFC3339, s); err == nil {
			return t.Format("2006-01-02"), nil
		}
		return s, fmt.Errorf("%q is not a date — expected YYYY-MM-DD", s)
	case FieldRef:
		ref := strings.TrimSpace(strings.Trim(s, "[]"))
		if !shortCodeRef.MatchString(ref) {
			return s, fmt.Errorf("%q is not a short code — expected E12, R2:E5 or RM1", s)
		}
		return ref, nil
	case FieldURL:
		u, err := url.Parse(s)
		if err != nil || (u.Scheme != "http" && u.Scheme != "https") || u.Host == "" {
			return s, fmt.Errorf("%q is not an http or https URL", s)
		}
		return s, nil
	}
	return s, nil
}

// shortCodeRef matches the reference shapes this product already understands.
var shortCodeRef = regexp.MustCompile(`^(?:R\d+:)?(?:E|S|SS|Q|T|RM|N)\d+(?::N\d+)?$|^R\d+$`)

// oneLineValue flattens newlines: a metadata value is a cell, and a value
// carrying a paragraph break is content wearing a field's clothes.
func oneLineValue(s string) string {
	s = strings.ReplaceAll(s, "\r\n", " ")
	s = strings.ReplaceAll(s, "\n", " ")
	s = strings.ReplaceAll(s, "\r", " ")
	return strings.Join(strings.Fields(s), " ")
}

// sortStrings keeps the domain package free of a sort import in one place only.
// sortStrings was an insertion sort, which is Θ(n²) on the input it actually
// gets: ValidateMetadata builds its key slice by ranging a map, so every call
// hands it a fresh random permutation. That was invisible while the only writer
// was an agent sending a dozen declared keys. A markdown file can carry a
// hundred thousand of them under the one-megabyte cap, and the preview route
// runs the validator before anything is refused — measured at over two minutes
// of CPU for a 1 MB file, from one request, on the caller's own section.
func sortStrings(s []string) {
	sort.Strings(s)
}

// MetadataIssues re-checks stored values against the declaration, so a reader
// can be told that `stage: "sent for review"` is not one of the options.
//
// It exists because MetadataReport is attached to a write and never stored: a
// value that was invalid on Tuesday is still invalid on Friday, and the person
// opening the document on Friday is the one who can fix it. Without this the
// client would have to re-implement enum membership, date parsing, URL schemes
// and the length cap as a second source of truth.
func MetadataIssues(specs []FieldSpec, values map[string]any) []MetadataIssue {
	var out []MetadataIssue
	for _, f := range specs {
		v, ok := values[f.Key]
		if !ok || v == nil {
			continue
		}
		if _, err := normalizeValue(f, v); err != nil {
			out = append(out, MetadataIssue{Key: f.Key, Reason: err.Error()})
		}
	}
	return out
}

// FieldTypeInfo describes one type to a client, so the editor does not carry a
// second copy of the catalogue that drifts on the seventh type.
type FieldTypeInfo struct {
	Type        FieldType `json:"type"`
	NeedsOption bool      `json:"needs_options"`
	Repeatable  bool      `json:"repeatable"`
	Hint        string    `json:"hint"`
}

// FieldSchemaInfo is everything a client needs to render a spec editor without
// hard-coding this package's rules.
type FieldSchemaInfo struct {
	Types        []FieldTypeInfo `json:"types"`
	ReservedKeys []string        `json:"reserved_keys"`
	Caps         map[string]any  `json:"caps"`
}

// FieldSchema serves the rules rather than restating them in the frontend. A
// cap the client believes and the server enforces will disagree exactly once,
// at the worst moment.
func FieldSchema() FieldSchemaInfo {
	return FieldSchemaInfo{
		Types: []FieldTypeInfo{
			{Type: FieldEnum, NeedsOption: true, Repeatable: true, Hint: "A short list of named options. The type that actually raises how often a field gets filled — prefer it wherever the answer can be enumerated."},
			{Type: FieldRef, Repeatable: true, Hint: "A short code of something in this product: E12, R2:E5, RM1."},
			{Type: FieldDate, Repeatable: false, Hint: "A calendar date, YYYY-MM-DD."},
			{Type: FieldText, Repeatable: true, Hint: "One short line. Anything needing a paragraph belongs in the document."},
			{Type: FieldNumber, Repeatable: false, Hint: "A number."},
			{Type: FieldURL, Repeatable: true, Hint: "An http or https address."},
		},
		ReservedKeys: ReservedFieldKeys(),
		Caps: map[string]any{
			"fields":      MaxSectionFields,
			"required":    MaxRequiredFields,
			"options":     MaxEnumOptions,
			"values":      MaxRepeatedValues,
			"text_max":    MaxFieldValueLen,
			"label_max":   MaxFieldLabelLen,
			"help_max":    MaxFieldHelpLen,
			"key_max":     MaxFieldKeyLen,
			"key_pattern": fieldKeyPattern.String(),
			// Import's caps ride here rather than in a payload of their own:
			// the drop hint, the refusal and the title counter all quote one of
			// them, and a number hard-coded in the frontend is a lie the day it
			// changes on the server.
			"import_max_bytes":  MaxImportFileBytes,
			"import_extensions": []string{".md", ".markdown"},
			// The section's writing instruction is edited on the same settings
			// card as the field spec, beside the fields it is supposed to name,
			// and its counter has to quote the number the server refuses on.
			"instruction_max": SectionInstructionMax,
		},
	}
}
